pkg/notify/notifier/pushover/message.go

```diff
@@ -0,0 +1,178 @@
+		"magic":        true,
+		"mechanical":   true,
+		"pianobar":     true,
+		"siren":        true,
+		"spacealarm":   true,
+		"tugboat":      true,
+		"alien":        true,
+		"climb":        true,
+		"persistent":   true,
+		"echo":         true,
+		"updown":       true,
+		"vibrate":      true,
+		"none":         true,
+	}
+)
+
+// Pushover message struct
+type pushoverMessage struct {
+	// required fields
+	// Token is a Pushover application API token, required.
+	Token string `json:"token"`
+	// UserKey is recipient's Pushover User Key, required.
+	UserKey string `json:"user"`
+	// Message is your text message, required.
+	Message string `json:"message"`
+
+	// common optional fields
+	// Attachment is an image to send with the message. NOTE: the caller has to close the reader.
+	Attachment io.Reader
+	// Device specifies a set of user's devices to send the message; all would be sent if empty
+	Device string `json:"device,omitempty"`
+	// Title is the message's title, otherwise application's name is used.
+	Title string `json:"title,omitempty"`
+	// Url is a supplementary URL to show with the message.
+	Url string `json:"url,omitempty"`
+	// UrlTitle is a title for Url, otherwise just the Url is shown.
+	UrlTitle string `json:"url_title,omitempty"`
+	// Priority ranges from -2 (no notification) to 2 (emergency, confirmation required).
+	Priority int `json:"priority,omitempty"`
+	// Sound is the name of one of the sounds supported by device clients.
+	Sound string `json:"sound,omitempty"`
+	// Timestamp is a Unix timestamp of your message's date and time to display to the user
+	Timestamp int64 `json:"timestamp,omitempty"`
+
+	// emergency Priority fields (no effects if Priority != 2)
+	Retry    time.Duration `json:"retry,omitempty"`
+	Expire   time.Duration `json:"expire,omitempty"`
+	Callback string        `json:"callback,omitempty"`
+
+	// styling fields (only one of both can be `1`, at most)
+	Html      int `json:"html,omitempty"`
+	Monospace int `json:"monospace,omitempty"`
+}
+
+func newPushoverMessage(token string, userKey string, message string) pushoverMessage {
+	return pushoverMessage{Token: token, UserKey: userKey, Message: message, Sound: "pushover"}
+}
+
+func (p *pushoverMessage) validate() (err error, warns []string) {
+	// Validate the application API token.
+	if !tokenRegex.MatchString(p.Token) {
+		err = fmt.Errorf("invalid Pushover application API token: %s", p.Token)
+		return
+	}
+
+	// Validate the User Key.
+	if !tokenRegex.MatchString(p.UserKey) {
+		err = fmt.Errorf("invalid Pushover User Key: %s", p.UserKey)
+		return
+	}
+
+	// Validate the length of the message.
+	msgLen := utf8.RuneCountInString(p.Message)
+	if msgLen <= 0 {
+		err = fmt.Errorf("message should not be empty")
+		return
+	}
+	if msgLen > MessageMaxLength {
+		warns = append(warns, fmt.Sprintf("the part of this message that exceeds %d runes is ignored", MessageMaxLength))
+	}
+
+	// Validate devices if configured.
+	if len(p.Device) > 0 {
```

Comment:
How to set the `device`? If the `device` can not be set, what is the value of this code? 
Similar to `Title`, `Url`, `UrlTitle`, `Priority`, `Sound`,  `Retry`, `Expire`.

Reply:
Now Devices, Sound and Title can be configured along with User Key. Please refer to https://github.com/kubesphere/notification-manager/blob/6229e77a79cb5cadc513d62513f96706f9879ebc/pkg/apis/v2beta2/receiver_types.go#L205

Other options are not used currently. If it is necessary to remove them, please let me know and I will do this in next commit.

Reply:
Please delete unused parameters.

Reply:
Resolved: https://github.com/txfs19260817/notification-manager/blob/16bc5ca72deefaf0d09abb41fa0bcc6a7a0bf852/pkg/notify/notifier/pushover/message.go#L49